mesonbuild/ast/interpreter.py

```diff
@@ -203,15 +204,15 @@ def func_subdir(self, node: BaseNode, args: T.List[TYPE_nvar], kwargs: T.Dict[st
             code = f.read()
         assert isinstance(code, str)
         try:
-            codeblock = mparser.Parser(code, absname).parse()
+            self.subdir_code_block = mparser.Parser(code, absname).parse()
         except mesonlib.MesonException as me:
             me.file = absname
             raise me
 
         self.subdir = subdir
         for i in self.visitors:
-            codeblock.accept(i)
-        self.evaluate_codeblock(codeblock)
+            self.subdir_code_block.accept(i)
+        self.evaluate_codeblock(self.subdir_code_block)
```

Comment:
I'm not sure this works, because that will call func_subdir() recursively. When evaluate_codeblock() returns self.subdir_code_block will have been changed to the latest subdir visited, not always the current one.

What about adding a method e.g. `self.visit_subdir(codeblock)` that does nothing in the base class but you override it in IntrospectionInterpreter?

Reply:
Or just build your dict of ast directly in the base class. We already have `self.processed_buildfiles.add(build_file)` change it from list to dict mapping to codeblock.